File: app/src/main/java/com/rjw/audioprofile/activity/MainActivity.java
package com.rjw.audioprofile.activity;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.PorterDuff;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.service.quicksettings.TileService;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.TextView;

import com.rjw.audioprofile.service.AudioProfileService;
import com.rjw.audioprofile.BuildConfig;
import com.rjw.audioprofile.service.QuickPanel;
import com.rjw.audioprofile.R;
import com.rjw.audioprofile.utils.Alerts;
import com.rjw.audioprofile.utils.AudioProfileList;
import com.rjw.audioprofile.utils.DisplayUtils;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

public class MainActivity extends AudioActivity {
    public final static String TAG = "AudioProfile";
    public final static int REQUEST_PERMISSIONS = 1;
    public final static int REQUEST_AUDIO_PROFILE = 2;
    public static final int ACTIVITY_SETTINGS = 3;
    public static final int ACTIVITY_SELECT_THEME_COLOUR = 4;

    public static final String CHANNEL_ID = "AudioProfileChannelId";
    public static final CharSequence CHANNEL_NAME = "AudioProfile";
    public static final String CHANNEL_DESCRIPTION = "AudioProfile";
    public static final int SERVICE_NOTIFICATION_ID = 100;

    public static final String PREF_APPLICATION_COLOUR = "ApplicationColour";

    private static MainActivity mThis;
    private static AudioProfileList mProfiles;
    private static NotificationManager mNm;
    private static int mAppColour;

    private RadioButton[] mRadioProfile = new RadioButton[AudioProfileList.NO_PROFILES];
    private ImageView[] mImageProfile = new ImageView[AudioProfileList.NO_PROFILES];
    private TextView[] mTextProfile = new TextView[AudioProfileList.NO_PROFILES];

    public static MainActivity getInstance() {
        if(mThis == null) {
            new MainActivity();
        }
        return mThis;
    }

    protected final ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(final ComponentName name, final IBinder service) {
        }

        @Override
        public void onServiceDisconnected(final ComponentName name) {
        }
    };

    @SuppressLint("MissingSuperCall")
    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mThis = this;

        // Setup handler for uncaught exceptions.
        Thread.setDefaultUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(final Thread thread, final Throwable e) {
                e.printStackTrace();
            }
        });

        // Load the settings.
        mProfiles = new AudioProfileList(this);

        mRadioProfile[0] = findViewById(R.id.radioProfile0);
        mRadioProfile[1] = findViewById(R.id.radioProfile1);
        mRadioProfile[2] = findViewById(R.id.radioProfile2);
        mRadioProfile[3] = findViewById(R.id.radioProfile3);
        mImageProfile[0] = findViewById(R.id.imageProfile0);
        mImageProfile[1] = findViewById(R.id.imageProfile1);
        mImageProfile[2] = findViewById(R.id.imageProfile2);
        mImageProfile[3] = findViewById(R.id.imageProfile3);
        mTextProfile[0] = findViewById(R.id.textProfile0);
        mTextProfile[1] = findViewById(R.id.textProfile1);
        mTextProfile[2] = findViewById(R.id.textProfile2);
        mTextProfile[3] = findViewById(R.id.textProfile3);

        // Set the profile names and icons.
        for(int profile = 0; profile < AudioProfileList.NO_PROFILES; profile++) {
            final AudioProfileList.AudioProfile audioProfile = AudioProfileList.getProfile(profile);
            mRadioProfile[profile].getButtonDrawable().setColorFilter(getConfigColour(), PorterDuff.Mode.SRC_ATOP);
            mTextProfile[profile].setText(audioProfile.name);
            mImageProfile[profile].setImageDrawable(AudioProfileList.getIcon(audioProfile.icon));
        }

        // Select the current profile.
        selectRadio(AudioProfileList.getCurrentProfile());

        // Check we have the required permissions.
        final Intent intent = new Intent(this, PermissionRequest.class);
        startActivityForResult(intent, REQUEST_PERMISSIONS);
        if(checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
               checkSelfPermission(Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[] { Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.ACCESS_BACKGROUND_LOCATION }, 0);
        }
        if(!isServiceRunning()) {
            startService();
        }
        final SharedPreferences prefs = getSharedPreferences(TAG, Activity.MODE_PRIVATE);
        mAppColour = prefs.getInt(PREF_APPLICATION_COLOUR, getColor(R.color.colourConfig));
        colourControls();
    }

    @Override
    protected void onActivityResult(final int requestCode, final int resultCode, final Intent data) {
        switch(requestCode) {
            case REQUEST_PERMISSIONS:
                updateTile(this);
                break;
            case REQUEST_AUDIO_PROFILE:
                if(resultCode == RESULT_OK) {
                    AudioProfileList.saveProfiles(this);
                    final int modified = data.getIntExtra(ProfileConfiguration.AUDIO_PROFILE, 0);
                    mTextProfile[modified].setText(AudioProfileList.getProfile(modified).name);
                    mImageProfile[modified].setImageDrawable(AudioProfileList.getIcon(AudioProfileList.getProfile(modified).icon));
                    if(AudioProfileList.getCurrentProfile() == modified) {
                        selectRadio(AudioProfileList.getCurrentProfile());
                    }
                }
                break;
            case ACTIVITY_SETTINGS:
                if(data != null) {
                    final int themeColour = data.getIntExtra(DisplayUtils.EXTRA_CUSTOM_COLOUR, mAppColour);
                    if(themeColour != mAppColour) {
                        mAppColour = themeColour;
                        setAppColour(mAppColour);
                    }
                }
                final Intent intent = new Intent(this, getClass());
                startActivity(intent);
                finish();
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
                break;
        }
    }

    public void onClickAudioProfile(final View v) {
        switch(v.getId()) {
            case R.id.radioProfile0:
            case R.id.imageProfile0:
            case R.id.textProfile0:
                selectRadio(0);
                break;
            case R.id.radioProfile1:
            case R.id.imageProfile1:
            case R.id.textProfile1:
                selectRadio(1);
                break;
            case R.id.radioProfile2:
            case R.id.imageProfile2:
            case R.id.textProfile2:
                selectRadio(2);
                break;
            case R.id.radioProfile3:
            case R.id.imageProfile3:
            case R.id.textProfile3:
                selectRadio(3);
                break;
        }
    }

    public void onClickAudioConfiguration(final View v) {
        final Intent intent = new Intent(this, ProfileConfiguration.class);
        switch(v.getId()) {
            case R.id.buttonConfigureProfile0:
                intent.putExtra(ProfileConfiguration.AUDIO_PROFILE, 0);
                break;
            case R.id.buttonConfigureProfile1:
                intent.putExtra(ProfileConfiguration.AUDIO_PROFILE, 1);
                break;
            case R.id.buttonConfigureProfile2:
                intent.putExtra(ProfileConfiguration.AUDIO_PROFILE, 2);
                break;
            case R.id.buttonConfigureProfile3:
                intent.putExtra(ProfileConfiguration.AUDIO_PROFILE, 3);
                break;
        }
        startActivityForResult(intent, REQUEST_AUDIO_PROFILE);
    }

    public void onClickAbout(final View view) {
        final StringBuilder about = new StringBuilder();
        try {
            final PackageInfo pInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            about.append(getString(R.string.about_version)).append(": ").append(pInfo.versionName);
            final Calendar buildDate = Calendar.getInstance();
            buildDate.setTimeInMillis(BuildConfig.TIMESTAMP);
            final StringBuilder builtDate = new StringBuilder(DateFormat.getDateInstance(DateFormat.SHORT).format(buildDate.getTime()));
            if(BuildConfig.DEBUG) {
                builtDate.append(" ").append(DateFormat.getTimeInstance(DateFormat.MEDIUM).format(buildDate.getTime()));
            }
            about.append("\n\n").append(getString(R.string.about_built)).append(" ").append(builtDate);
            about.append("\n\n").append(String.format(getString(R.string.copyright), new SimpleDateFormat("yyyy", Locale.getDefault()).format(buildDate.getTime())));
            final StringBuilder title = new StringBuilder(getString(R.string.about_title)).append(" ").append(getString(R.string.app_name));
            Alerts.alert(title, about);
        } catch(final Throwable e) {
            // Do nothing.
        }
    }

    public void onClickSettings(final View v) {
        final Intent intent = new Intent(this, Settings.class);
        startActivityForResult(intent, ACTIVITY_SETTINGS);
    }

    public void onClickClose(final View v) {
        finish();
    }

    public void selectRadio(final int profile) {
        AudioProfileList.setCurrentProfile(profile);
        switch(profile) {
            case 0:
                mRadioProfile[0].setChecked(true);
                mRadioProfile[1].setChecked(false);
                mRadioProfile[2].setChecked(false);
                mRadioProfile[3].setChecked(false);
                break;
            case 1:
                mRadioProfile[0].setChecked(false);
                mRadioProfile[1].setChecked(true);
                mRadioProfile[2].setChecked(false);
                mRadioProfile[3].setChecked(false);
                break;
            case 2:
                mRadioProfile[0].setChecked(false);
                mRadioProfile[1].setChecked(false);
                mRadioProfile[2].setChecked(true);
                mRadioProfile[3].setChecked(false);
                break;
            case 3:
                mRadioProfile[0].setChecked(false);
                mRadioProfile[1].setChecked(false);
                mRadioProfile[2].setChecked(false);
                mRadioProfile[3].setChecked(true);
                break;
        }
        AudioProfileList.applyProfile(this);
        updateTile(this);
    }

    public static void updateTile(final Context context) {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            try {
                TileService.requestListeningState(context, new ComponentName(context, QuickPanel.class));
            } catch(Exception e) {
                Log.d("AudioProfile", "Tile update failure - starting service");
            }
        }
    }

    public static AudioProfileList getProfiles() {
        return mProfiles;
    }

    public static int getConfigColour() {
        return mAppColour;
    }

    public static int getWhiteColour() {
        return mThis.getColor(R.color.colourWhiteText);
    }

    public void startService() {
        final Intent serviceIntent = new Intent(mThis, AudioProfileService.class);
        try {
            bindService(serviceIntent, mConnection, Context.BIND_AUTO_CREATE);
            startService(serviceIntent);
        } catch(Exception e) {
            Alerts.toast(e.toString());
        }
    }

    private boolean isServiceRunning() {
        try {
            final ActivityManager manager = (ActivityManager)getSystemService(Context.ACTIVITY_SERVICE);
            for(final ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
                if(AudioProfileService.class.getName().equals(service.service.getClassName())) {
                    return true;
                }
            }
        } catch(Exception e) {
            // Do nothing.
        }
        return false;
    }

    public static void createNotificationChannel(final Context context) {
        try {
            mNm = (NotificationManager)mThis.getSystemService(Context.NOTIFICATION_SERVICE);
            if(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
                final NotificationChannel channel = new NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_MIN);
                channel.setDescription(CHANNEL_DESCRIPTION);
                channel.setShowBadge(false);
                channel.enableLights(false);
                channel.setSound(null, null);
                mNm.createNotificationChannel(channel);
            }
        } catch(Exception e) {
            Alerts.toast("Creating notification channel: " + e.getMessage());
        }
    }

    public static void showServiceNotification(final Service service, final String msg, final PendingIntent pendingIntent) {
        if(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            try {
                final Notification.Builder builder;
                builder = new Notification.Builder(service)
                        .setChannelId(CHANNEL_ID)
                        .setSmallIcon(R.drawable.notification)
                        .setContentText(msg)
                        .setContentIntent(pendingIntent);
                service.startForeground(SERVICE_NOTIFICATION_ID, builder.build());
            } catch(Exception e) {
                Alerts.toast("Creating notification: " + e.getMessage());
            }
            Log.d(TAG, msg);
        }
    }

    public static void setAppColour(final int colour) {
        mAppColour = colour;
        final SharedPreferences prefs = mThis.getSharedPreferences(TAG, Activity.MODE_PRIVATE);
        prefs.edit().putInt(PREF_APPLICATION_COLOUR, mAppColour).apply();
    }
}

File: app/src/main/java/com/rjw/audioprofile/activity/MainActivity.kt
package com.rjw.audioprofile.activity

import android.Manifest
import android.annotation.SuppressLint
import android.app.ActivityManager
import android.app.NotificationManager
import android.app.WallpaperManager
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.os.PowerManager
import android.service.quicksettings.TileService
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ImageView
import android.widget.RadioButton
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.view.isVisible
import com.rjw.audioprofile.R
import com.rjw.audioprofile.databinding.ActivityMainBinding
import com.rjw.audioprofile.databinding.ContentMainBinding
import com.rjw.audioprofile.service.AudioProfileService
import com.rjw.audioprofile.service.Notifications
import com.rjw.audioprofile.service.QuickPanel
import com.rjw.audioprofile.utils.Alerts
import com.rjw.audioprofile.utils.AudioProfileList
import com.rjw.audioprofile.utils.CLEAR_LOG
import com.rjw.audioprofile.utils.MinutesAdapter
import com.rjw.audioprofile.utils.Mode
import com.rjw.audioprofile.utils.ProfileAdapter
import com.rjw.audioprofile.utils.TAG
import com.rjw.audioprofile.utils.setColorFilter
import java.util.Calendar

@Suppress("DEPRECATION")
class MainActivity : AudioActivity() {
    private val REQUEST_PERMISSION_RESPONSE = 1
    private lateinit var binding: ActivityMainBinding
    private lateinit var bindingContent: ContentMainBinding
    private val lockTimings = arrayOf(1, 2, 5, 10, 20, 30, 60, 90, 120)
    private val radioProfile by lazy { arrayOfNulls<RadioButton>(AudioProfileList.noProfiles) }
    private val imageProfile by lazy { arrayOfNulls<ImageView>(AudioProfileList.noProfiles) }
    private val textProfile by lazy { arrayOfNulls<TextView>(AudioProfileList.noProfiles) }
    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {}
        override fun onServiceDisconnected(name: ComponentName?) {}
    }
    private var profileLockChanged = false
    private val requestPermissions = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { _ ->
        updateTile()
    }
    private val requestAudioProfile = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if(result.resultCode == RESULT_OK) {
            result.data?.let { data ->
                AudioProfileList.saveProfiles(this)
                val modified = data.getIntExtra(ProfileConfiguration.AUDIO_PROFILE, 0)
                textProfile[modified]?.text = AudioProfileList.getProfile(modified).name
                imageProfile[modified]?.setImageDrawable(AudioProfileList.getIcon(AudioProfileList.getProfile(modified).icon))
                updateControls()
            }
        }
    }
    private var askedPermissions = false

    /**
     * Create the activity.
     * @param savedInstanceState The state information for the activity.
     */
    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        windowRatio = floatArrayOf(0.9f, 0.7f)
        setContentView(R.layout.activity_main)
        binding = ActivityMainBinding.bind(view)
        bindingContent = ContentMainBinding.bind(binding.layoutMain)

        // Setup handler for uncaught exceptions.
        Thread.setDefaultUncaughtExceptionHandler { _, e -> e.printStackTrace() }

        // Load the settings.
        AudioProfileList.initialise(this)
        radioProfile[0] = bindingContent.radioProfile0
        radioProfile[1] = bindingContent.radioProfile1
        radioProfile[2] = bindingContent.radioProfile2
        radioProfile[3] = bindingContent.radioProfile3
        imageProfile[0] = bindingContent.imageProfile0
        imageProfile[1] = bindingContent.imageProfile1
        imageProfile[2] = bindingContent.imageProfile2
        imageProfile[3] = bindingContent.imageProfile3
        textProfile[0] = bindingContent.textProfile0
        textProfile[1] = bindingContent.textProfile1
        textProfile[2] = bindingContent.textProfile2
        textProfile[3] = bindingContent.textProfile3

        if(!isServiceRunning) {
            startService()
        }

        // Check we have the required additional permissions.
        val doNotDisturb = (getSystemService(NOTIFICATION_SERVICE) as NotificationManager).isNotificationPolicyAccessGranted
        val batteryOptimizations = (getSystemService(POWER_SERVICE) as PowerManager).isIgnoringBatteryOptimizations(packageName)
        val notifications =
            !(Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED)
        if(!doNotDisturb || !notifications || !batteryOptimizations) {
            val intent = Intent(this, PermissionRequest::class.java)
            requestPermissions.launch(intent)
        }

        // Set the profile names and icons.
        for(profile in 0 until AudioProfileList.noProfiles) {
            val audioProfile = AudioProfileList.getProfile(profile)
            radioProfile[profile]?.buttonDrawable?.setColorFilter(configColour, Mode.SRC_ATOP)
            textProfile[profile]?.text = audioProfile.name
            imageProfile[profile]?.setImageDrawable(AudioProfileList.getIcon(audioProfile.icon))
        }

        // Set up the lock adapters.
        val profileLockTime = AudioProfileList.lockProfileTime
        val lockAdapter = MinutesAdapter(this, lockTimings)
        bindingContent.spinnerLockProfile.adapter = lockAdapter
        bindingContent.spinnerLockProfile.background.setColorFilter(configColour, Mode.SRC_ATOP)
        bindingContent.checkboxLockProfile.buttonDrawable?.setColorFilter(configColour, Mode.SRC_ATOP)

        // Set up the profile lock time.
        var profileLocked = AudioProfileList.profileLocked
        if(profileLocked and (Calendar.getInstance().timeInMillis > AudioProfileList.profileLockStartTime + profileLockTime * 60000)) {
            profileLocked = false
        }
        bindingContent.checkboxLockProfile.isChecked = profileLocked

        // Set the handler for the lock time spinner.
        AdapterView.OnItemClickListener { _, _, _, _ -> profileLockChanged = true }

        // Set up the control operations.
        bindingContent.about.setOnClickListener {
            startActivity(Intent(this, AboutActivity::class.java))
        }
        bindingContent.radioProfile0.setOnClickListener { setAudioProfile(0) }
        bindingContent.imageProfile0.setOnClickListener { setAudioProfile(0) }
        bindingContent.textProfile0.setOnClickListener { setAudioProfile(0) }
        bindingContent.rowProfile0.setOnClickListener { setAudioProfile(0) }
        bindingContent.buttonConfigureProfile0.setOnClickListener { configureAudioProfile(0) }
        bindingContent.radioProfile1.setOnClickListener { setAudioProfile(1) }
        bindingContent.imageProfile1.setOnClickListener { setAudioProfile(1) }
        bindingContent.textProfile1.setOnClickListener { setAudioProfile(1) }
        bindingContent.rowProfile1.setOnClickListener { setAudioProfile(1) }
        bindingContent.buttonConfigureProfile1.setOnClickListener { configureAudioProfile(1) }
        bindingContent.radioProfile2.setOnClickListener { setAudioProfile(2) }
        bindingContent.imageProfile2.setOnClickListener { setAudioProfile(2) }
        bindingContent.textProfile2.setOnClickListener { setAudioProfile(2) }
        bindingContent.rowProfile2.setOnClickListener { setAudioProfile(2) }
        bindingContent.buttonConfigureProfile2.setOnClickListener { configureAudioProfile(2) }
        bindingContent.radioProfile3.setOnClickListener { setAudioProfile(3) }
        bindingContent.imageProfile3.setOnClickListener { setAudioProfile(3) }
        bindingContent.textProfile3.setOnClickListener { setAudioProfile(3) }
        bindingContent.rowProfile3.setOnClickListener { setAudioProfile(3) }
        bindingContent.buttonConfigureProfile3.setOnClickListener { configureAudioProfile(3) }
        bindingContent.checkboxLockProfile.setOnClickListener {
            AudioProfileList.profileLocked = bindingContent.checkboxLockProfile.isChecked
            profileLockChanged = true
            updateTile()
        }
        bindingContent.textSettings.setOnClickListener {
            bindingContent.layoutSettings.visibility = if(bindingContent.layoutSettings.isVisible) View.GONE else View.VISIBLE
        }
        bindingContent.imageSettings.setOnClickListener {
            bindingContent.layoutSettings.visibility = if(bindingContent.layoutSettings.isVisible) View.GONE else View.VISIBLE
        }
        bindingContent.checkboxEnterWifiDefault.setOnClickListener {
            AudioProfileList.enterWifiProfile =
                if(bindingContent.checkboxEnterWifiDefault.isChecked) -1 else bindingContent.spinnerEnterWifi.selectedItemPosition
            updateControls()
        }
        bindingContent.spinnerEnterWifi.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, p2: Int, p3: Long) {
                AudioProfileList.enterWifiProfile = bindingContent.spinnerEnterWifi.selectedItemPosition
            }

            override fun onNothingSelected(p0: AdapterView<*>?) {}
        }
        bindingContent.checkboxExitWifiDefault.setOnClickListener {
            AudioProfileList.exitWifiProfile =
                if(bindingContent.checkboxExitWifiDefault.isChecked) -1 else bindingContent.spinnerExitWifi.selectedItemPosition
            updateControls()
        }
        bindingContent.spinnerExitWifi.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, p2: Int, p3: Long) {
                AudioProfileList.exitWifiProfile = bindingContent.spinnerExitWifi.selectedItemPosition
            }

            override fun onNothingSelected(p0: AdapterView<*>?) {}
        }
        bindingContent.buttonClose.setOnClickListener {
            AudioProfileList.enterWifiProfile =
                if(bindingContent.checkboxEnterWifiDefault.isChecked) -1 else bindingContent.spinnerEnterWifi.selectedItemPosition
            AudioProfileList.exitWifiProfile =
                if(bindingContent.checkboxExitWifiDefault.isChecked) -1 else bindingContent.spinnerExitWifi.selectedItemPosition
            AudioProfileList.lockProfileTime =
                bindingContent.spinnerLockProfile.getItemAtPosition(bindingContent.spinnerLockProfile.selectedItemPosition) as Int
            if(profileLockChanged) {
                AudioProfileList.profileLocked = bindingContent.checkboxLockProfile.isChecked
                AudioProfileList.profileLockStartTime = Calendar.getInstance().timeInMillis
            }
            finish()
        }

        colourControls()
        if(savedInstanceState == null) {
            setAudioProfile(AudioProfileList.currentProfile)
        }
        if(intent.hasExtra(CLEAR_LOG)) {
            Alerts.clearLog()
        }
    }

    /**
     * Resume the activity.
     */
    @RequiresApi(Build.VERSION_CODES.P)
    override fun onResume() {
        super.onResume()
        // Request standard application permissions.
        if(!askedPermissions) {
            if(checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), REQUEST_PERMISSION_RESPONSE)
            } else {
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && checkSelfPermission(Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    ActivityCompat.requestPermissions(
                        this,
                        arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION),
                        REQUEST_PERMISSION_RESPONSE
                    )
                }
            }
            askedPermissions = true
        }
    }

    /**
     * Deal with permission requests.
     * @param requestCode  The code for the permission request.
     * @param permissions  The permissions being requested.
     * @param grantResults The results for each permission.
     */
    @Deprecated("Deprecated in Java")
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    /**
     * Select the radio button and the current profile.
     * @param profile The profile to be set.
     */
    private fun setAudioProfile(profile: Int) {
        AudioProfileList.currentProfile = profile
        when(profile) {
            0 -> {
                radioProfile[0]?.isChecked = true
                radioProfile[1]?.isChecked = false
                radioProfile[2]?.isChecked = false
                radioProfile[3]?.isChecked = false
            }
            1 -> {
                radioProfile[0]?.isChecked = false
                radioProfile[1]?.isChecked = true
                radioProfile[2]?.isChecked = false
                radioProfile[3]?.isChecked = false
            }
            2 -> {
                radioProfile[0]?.isChecked = false
                radioProfile[1]?.isChecked = false
                radioProfile[2]?.isChecked = true
                radioProfile[3]?.isChecked = false
            }
            3 -> {
                radioProfile[0]?.isChecked = false
                radioProfile[1]?.isChecked = false
                radioProfile[2]?.isChecked = false
                radioProfile[3]?.isChecked = true
            }
        }
        AudioProfileList.currentProfile = profile
        AudioProfileList.profileLocked = false
    }

    /**
     * Update the specified audio profile.
     * @param profile The profile to be configured.
     */
    private fun configureAudioProfile(profile: Int) {
        val intent = Intent(this, ProfileConfiguration::class.java)
        intent.putExtra(ProfileConfiguration.AUDIO_PROFILE, profile)
        requestAudioProfile.launch(intent)
    }

    /**
     * Update the controls to reflect the current settings.
     */
    private fun updateControls() {
        // Now update the spinners for the entry and exit events.
        val adapterLockTime = bindingContent.spinnerLockProfile.adapter
        for(item in 0 until adapterLockTime.count) {
            if(adapterLockTime.getItem(item) == AudioProfileList.lockProfileTime) {
                bindingContent.spinnerLockProfile.setSelection(item)
                break
            }
        }
        bindingContent.checkboxLockProfile.isChecked = AudioProfileList.profileLocked
        bindingContent.spinnerEnterWifi.visibility = if(bindingContent.checkboxEnterWifiDefault.isChecked) View.INVISIBLE else View.VISIBLE
        bindingContent.spinnerExitWifi.visibility = if(bindingContent.checkboxExitWifiDefault.isChecked) View.INVISIBLE else View.VISIBLE
        bindingContent.spinnerLockProfile.visibility = if(bindingContent.checkboxLockProfile.isChecked) View.VISIBLE else View.INVISIBLE

        val adapter = ProfileAdapter(this, AudioProfileList.getProfiles().toTypedArray())
        bindingContent.spinnerEnterWifi.adapter = adapter
        bindingContent.checkboxEnterWifiDefault.buttonDrawable?.setColorFilter(configColour, Mode.SRC_ATOP)
        bindingContent.spinnerExitWifi.adapter = adapter
        bindingContent.checkboxExitWifiDefault.buttonDrawable?.setColorFilter(configColour, Mode.SRC_ATOP)
        var profile = AudioProfileList.enterWifiProfile
        if(profile == -1) {
            bindingContent.checkboxEnterWifiDefault.isChecked = true
        } else {
            bindingContent.spinnerEnterWifi.setSelection(profile)
        }
        profile = AudioProfileList.exitWifiProfile
        if(profile == -1) {
            bindingContent.checkboxExitWifiDefault.isChecked = true
        } else {
            bindingContent.spinnerExitWifi.setSelection(profile)
        }
        colourControls()
    }

    /**
     * Start the audio listener service.
     */
    private fun startService() {
        val serviceIntent = Intent(mThis, AudioProfileService::class.java)
        try {
            bindService(serviceIntent, connection, BIND_AUTO_CREATE)
            startService(serviceIntent)
        } catch(_: Exception) {
            // Do nothing...
        }
    }

    private val isServiceRunning: Boolean
        /**
         * Return if the service is running.
         */
        get() {
            try {
                val am = getSystemService(ACTIVITY_SERVICE) as ActivityManager?
                if(am != null) {
                    for(service in am.getRunningServices(Int.MAX_VALUE)) {
                        if(AudioProfileService::class.java.name == service.service.className) {
                            return true
                        }
                    }
                }
            } catch(_: Exception) {
                // Do nothing.
            }
            return false
        }

    @SuppressLint("StaticFieldLeak")
    companion object {
        private var mThis: MainActivity? = null

        val instance: MainActivity?
            /**
             * Return an instance of the main activity.
             */
            get() {
                if(mThis == null) {
                    MainActivity()
                }
                return mThis
            }

        /**
         * Update the quick panel tile to reflect the current profile.
         */
        fun updateTile() {
            try {
                mThis?.let { instance ->
                    instance.updateControls()
                    TileService.requestListeningState(instance, ComponentName(instance.applicationContext, QuickPanel::class.java))
                    Notifications.updateNotification(instance)
                }
            } catch(_: Exception) {
                // Do nothing.
            }
        }

        val configColour: Int
            /**
             * Get the current application colour.
             */
            get() {
                return when {
                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.S -> {
                        mThis?.getColor(R.color.colourConfig) ?: 0
                    }
                    else -> {
                        val wm = WallpaperManager.getInstance(mThis)
                        wm.getWallpaperColors(WallpaperManager.FLAG_SYSTEM)?.primaryColor?.toArgb() ?: 0
                    }
                }
            }

        val secondaryColour: Int
            /**
             * Get the current application secondary colour.
             */
            get() {
                return when {
                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.S -> {
                        mThis?.getColor(R.color.colourAccent) ?: 0
                    }
                    else -> {
                        val wm = WallpaperManager.getInstance(mThis)
                        wm.getWallpaperColors(WallpaperManager.FLAG_SYSTEM)?.secondaryColor?.toArgb() ?: 0
                    }
                }
            }

        val whiteColour: Int
            /**
             * Get the white colour.
             */
            get() = mThis?.getColor(R.color.colourWhiteText) ?: 0xFFFFFF
    }

    init {
        mThis = this
    }
}